core/src/main/java/io/questdb/cairo/sql/Function.java

```diff
@@ -252,6 +266,49 @@ default boolean isUndefined() {
         return getType() == ColumnType.UNDEFINED;
     }
 
+    /**
+     * DeepClone does not clone the running-states, only the init states of the function,
+     * must be called before Function::init.
+     * Used for parallel filter/groupBy execution, so that avoid Function::parser for every worker.
+     * Note: CursorFunction and WindowFunction are not supported now.
+     *
+     * @return deep clones of the function
+     */
+    default Function deepClone() {
+        try {
+            Class<?> cls = this.getClass();
+            Constructor<?> funcCons = cls.getDeclaredConstructors()[0];
+            funcCons.setAccessible(true);
+            int parameterCount = funcCons.getParameterCount();
+            Object[] pArgs = new Object[parameterCount];
+            for (int i = 0; i < parameterCount; i++) {
+                pArgs[i] = DEFAULT_PRIMITIVE_VALUES.get(funcCons.getParameterTypes()[i]);
```

Comment:
In many cases `Function` constructor accepts another function, e.g.
```java
public SubStringFunc(Function strFunc, Function startFunc, Function lenFunc)
```

We should handle that case and provide a stub `Function` object to the constructor instead of the current `null`. `null` is very NPE-prone, so we should give up if we don't know the object argument type instead of providing a `null`.

Also, `SubStringFunc` is an interesting function since it initializes its state in the constructor based on the argument functions. This would break `deepClone()`, so all such functions need to be patched to initialize the state in `init()`.